File: C++ Space Shooter src/player.cpp
#include "player.h"
#include "main.h"
#include "game.h"
#include "utils.h"

#include "raylib.h"
#include "raymath.h"

#include <string>

void Player::GetInputs()
{
	input.x = (float)IsKeyDown(KEY_D) - IsKeyDown(KEY_A);
	input.y = (float)IsKeyDown(KEY_S) - IsKeyDown(KEY_W);
	input = Vector2Normalize(input);

	rotation = GetAngleTo(pos, ScreenToWorld(GetMousePosition()));

	if (canShoot && IsMouseButtonDown(MOUSE_BUTTON_LEFT))
	{
		canShoot = false;
		shootTimer.StartTimer();

		CreateLaser(pos, 25.0f, rotation, 3);
	}
}

void Player::Cooldowns()
{
	double currentTime = GetTime();

	if (!canShoot && shootTimer.TimerDone())
	{
		canShoot = true;
	}
}

void Player::Update(float delta)
{
	GetInputs();

	// Movement Calculations
	accelerationVector = Vector2Multiply(input, { acceleration, acceleration });
	accelerationVector = Vector2Subtract(accelerationVector, Vector2Multiply(velocity, { drag, drag }));

	velocity = Vector2Add(velocity, Vector2Multiply(accelerationVector, { delta, delta }));

	pos = Vector2Add(pos, Vector2Multiply(velocity, { delta, delta }));

	Cooldowns();
}

void Player::Draw()
{
	image.Draw(pos, rotation);
}

File: 2022-2023/Space Shooter/data/modules/player.py
import pygame

from data.modules.image import Image
from data.modules.utils import get_angle_to


class Player:
	acceleration = 1.3
	drag = 0.12

	def __init__(self, pos):
		self.pos = pygame.Vector2(pos)

		self.input = pygame.Vector2()

		self.acceleration = pygame.Vector2()
		self.movement = pygame.Vector2()

		self.image = Image("player", scale=6)

		self.alive = True
		self.health = 100
		self.radius = 20

	def damage(self, damage):
		self.health -= damage

		if self.health <= 0:
			self.alive = False
			return True
		return False

	def get_inputs(self, scroll):
		keys_pressed = pygame.key.get_pressed()

		self.input.x = keys_pressed[pygame.K_d] - keys_pressed[pygame.K_a]
		self.input.y = keys_pressed[pygame.K_s] - keys_pressed[pygame.K_w]

		self.image.angle = get_angle_to(self.pos - scroll, pygame.mouse.get_pos())

	def move(self, delta):
		if self.input.length() != 0:
			self.acceleration = self.input.normalize() * Player.acceleration
		else:
			self.acceleration.xy = 0, 0

		self.acceleration -= Player.drag * self.movement

		self.movement += self.acceleration * delta

		self.pos += self.movement * delta

	def update(self, delta, scroll):
		self.get_inputs(scroll)

		self.move(delta)

	def draw(self, surface, scroll):
		self.image.draw(surface, self.pos - scroll)
